keps/sig-node/20190517-image-decryption.md

```diff
@@ -0,0 +1,330 @@
+authors:
+  - "@harche"
+owning-sig: sig-node
+participating-sigs:
+  - sig-architecture
+reviewers:
+  - smarterclayton
+  - tallclair
+  - yujuhong
+approvers:
+  - smarterclayton
+  - tallclair
+creation-date: 2019-05-16
+status: provisional
+---
+
+# Adding the Support for Encrypted Images
+
+## Table of Contents
+
+* [Summary](#summary)
+* [Motivation](#motivation)
+  * [Goals](#goals)
+  * [Non\-Goals](#non-goals)
+  * [User Stories](#user-stories)
+* [Proposal](#proposal)
+  * [API](#api)
+    * [Image Handler](#Image-handler)
+  * [Relationship with imagePullPolicy](#Relationship-with-imagePullPolicy)
+  * [Consumption of the ImageDecryptSecrets](#Consumption-of-the-ImageDecryptSecrets)
+* [Alternatives Considered](#alternatives-considered)
+* [Graduation Criteria](#graduation-criteria)
+* [Implementation History](#implementation-history)
+
+## Summary
+
+The underlying specification for the containers, the OCI spec, is soon going to support encrypted images. Kubernetes should be able to support decryption of these encrypted images with the addition of a new type of `Secret`, which we would like to call `ImageDecryptSecret`. 
+
+Along with OCI spec, there is an ongoing effort to enable the support for encrypted images in containerd. 
+
+OCI Spec Issue - https://github.com/opencontainers/image-spec/issues/747
+
+OCI Spec PR - https://github.com/opencontainers/image-spec/pull/775
+
+POC - https://github.com/harche/kubernetes/commit/8a0ef9898a55110d7e41f8e1a846c66cfc2fa691
+
+POC Demo - https://youtu.be/S3FK4y5McOk
+
+
+## Motivation
+
+The kubernetes worker nodes are where container images are pulled by a runtime such as `containerd`. If the images pulled are encrypted then containerd will have to decrypt them before running the pod. In order to be able to decrypt the images, the worker node needs to have access to corresponding the private keys.
+
+Kubernetes `Secrets` are used to securely deliver sensitive data to pods in corresponding worker nodes. We need to have a secret that can be utilized *before* the pod is provisioned. Regular Kubernetes secret gets attached to the pod as tmpfs mount after the pod is started. However, there exists another type of kubernetes secret called `ImagePullSecrets`. ImagePullSecrets are used to pull the images from the private container image registry, hence they contain login credentials. `ImagePullSecrets` get utilized *before* the pod is started, this is exactly the same kind of requirement for being able to decrypt the encrypted container image. We need to have a `secret` that can be used *before* the pod is provisioned to decrypt the image. Hence, we are submitting this KEP to kubernetes community to propose a new type of kubernetes secret that is modeled after `ImagePullSecret`, called `ImageDecryptSecret`. While the `ImagePullSecret` holds the login credentials for the private registry, the `ImageDecryptSecret` will hold the private keys to decrypt encrypted images.
```

Comment:
Are the private key(s) the same used to encrypt the images? I'm assuming the authors of the KEP want to keep it non-opinionated and the onus will be on the user/operator to use this feature in whatever way they like, right?

Reply:
@pires Yes, they have to be the same; the public key was used during encryption, its private key is needed during decryption and without it the image won't be decryptable.